File: 074_SearchA2DMatrix.py
#84.3%
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        m = len(matrix)
        if m == 0:
            return False
        n = len(matrix[0])
        if n == 0:
            return False
        
        left, right = 0, m * n - 1
        while left <= right:
            mid = (left + right) / 2
            rowIndex = mid / n
            colIndex = mid - rowIndex * n
            if matrix[rowIndex][colIndex] > target:
                right = mid - 1
            elif matrix[rowIndex][colIndex] < target:
                left = mid + 1
            else:
                return True
        return False

File: 074_SearchA2DMatrix.cpp
//22.5%
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m(matrix.size());
		if(m == 0)
			return false;
		int n(matrix[0].size());
		if(n == 0)
			return false;
		
		int left(0), right(m * n - 1);
		while (left <= right){
			int mid((left + right) / 2);
			int rowIndex(mid / n), colIndex(mid  - rowIndex * n);
			if(matrix[rowIndex][colIndex] > target)
				right = mid - 1;
			else if(matrix[rowIndex][colIndex] < target)
				left = mid + 1;
			else
				return true;
		}
		return false;
    }
};